// @dethcrypto/eth-sdk/dist/peripherals/fs.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.realFs = void 0;
const fs_1 = __importDefault(require("fs"));
const fs_extra_1 = __importDefault(require("fs-extra"));
const glob_1 = __importDefault(require("glob"));
const mkdirp_1 = __importDefault(require("mkdirp"));
const tmp_promise_1 = require("tmp-promise");
const util_1 = require("util");
exports.realFs = {
    exists: fs_1.default.existsSync,
    write: (p, c) => fs_extra_1.default.writeFile(p, c),
    ensureDir: async (p) => {
        await (0, mkdirp_1.default)(p);
    },
    copy: (src, dest) => fs_extra_1.default.copy(src, dest),
    readDir: (path) => fs_extra_1.default.readdir(path),
    tmpDir: async (prefix) => {
        const { path } = await (0, tmp_promise_1.dir)({ prefix });
        return path;
    },
    glob: (0, util_1.promisify)(glob_1.default),
};
